### PR TITLE
Add debug functionality

### LINKED ISSUE #2: Built-in debug info
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/emittery/issues/2)

Running `DEBUG=emittery node x.js` or `DEBUG=* node x.js` should output debug information/trace of events.

Events can be hard to debug, so this is important.

I'm looking for feedback on what the trace output should look like.

---

*I don't want to depend on the `debug` module, but we could easily detect the environment variable manually.*

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Submitted pull Requests
- [#53 Implements debug feature as described in #2](https://issuehunt.io/r/sindresorhus/emittery/pull/53)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/emittery/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by sbencoding**
This PR aims to implement a built-in debug feature described in #2
#### How it's implemented
There's a static function in the `Emittery` class called `debugLogger`, that takes in 4 parameters
* type - type of debug log
* debugName - set by the instance of Emittery that generated this log
* eventName - event name related to the log
* eventData - in case of emit, this is the data sent by it

Debugging can be triggered in 3 ways currently:
1. the `DEBUG` environment variable is set to *emittery* or *
2. the static `isDebug` function in Emittery returns true (by default this handles the environment variable, but this function can be overridden, to enable package wide debugging instead of the 3rd method.
3. the `isDebug` variable is set to true on one or more instances.
#### Questions/suggestions
I wasn't sure if the default logger function should include some sort of timestamp.

It might add too much complexity but a verbosity option could be nice. For example currently every emit, subscribe, unsubscribe etc. event is logged, but in some cases they might not all be interesting. Of course this can be achieved by overriding the `debugLogger` function on `Emittery`, but it would be a nice convenience option.

Is it acceptable that there's one static logging function on `Emittery` or should the individual instanced of it be able to override this function. This would add the ability for different instances of `Emittery` to log differently based on their function in an application.

Fixes #2

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#2: Built-in debug info](https://issuehunt.io/repos/112248999/issues/2)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by sbencoding** [comment]
Not sure why codecov doesn't pickup that I test the `debugLogger` static function on `Emittery`.
Unfortunately I'm not familiar with this software, but I tried calling the method directly from one of the test cases and it still showed that I missed that function :(

File: index.js [hunk 1]
```diff
@@ -181,16 +181,37 @@ class Emittery {
 		};
 	}
 
-	constructor() {
+	static isDebug() {
```
Comment (sindresorhus) [inline index.js:184]: This should be a getter.

File: index.js [hunk 2]
```diff
@@ -181,16 +181,37 @@ class Emittery {
...
+		return process.env.DEBUG === 'emittery' || process.env.DEBUG === '*';
+	}
+
+	static debugLogger(type, debugName, eventName, eventData) {
+		if (typeof eventName !== 'string') {
+			eventName = eventName.toString();
```
Comment (sindresorhus) [inline index.js:190]: This is moot. Template interpolation already calls toString.

**Comment by sindresorhus** [comment]
> Questions/suggestions

Share your opinion on these questions.

**Comment by sbencoding** [comment]
I've implemented the requested changes, updated documentation and tests where it was needed.
My thoughts on the questions:
#### Timestamps
I think it would be useful only in specific scenarios if the default logging function printed out timestamps as well. The use case I can currently think of where time is relevant is an application where an emit happens periodically e.g. every 10 seconds. I think a `minutes:seconds:milliseconds` format would be the best in this case, since I don't expect the debug to run for hours or days.
Because I can't think of better/more use cases I don't think this would be useful to have it in by default.
#### Verbosity filter
I think this is useful because with multiple events and multiple instances of Emittery the logs could get cluttered very quickly if we log every subscribe/unsubscribe event too. I can think of two approaches that would be good here.
1. Verbosity levels from 1 to 3
Where 1 only logs emits, 2 logs emits and sub/unsub, and 3 logs everything
2. An array of strings that defines the event types we're interested in
For example we could add emit, subscribe and clear to it, if those are the events we're interested in. This allows for more precise control than option number 1 but also takes a bit more time to set up.

The other thing that needs to be decided here is where we implement the verbosity setting, in the Emittery class, so it affects all instances at once, or on a per instance level, so it's more precisely controllable. However on the per instance level we would have to manually update every instance if we wanted to change the setting for all instances.
Overall I would say that implementing it globally is better because it would be more efficient at reducing the overall clutter.

#### Static logger function
I think it would be better to have this on a per instance basis, but I guess this question is kind of the same as the implementation of the verbosity filter. The benefit of logging on a per instance basis would be that the different instances could access potentially useful data that is only available in their context. But again the drawback is that we lose the ability to update the logging function of all instances all at once.
Overall I think implementing this on a per instance basis would provide more functionality than it would take away. I couldn't come up with a use case where updating every logger function would be that important, but I can imagine that some instances would like to log information only available in their context.

**Comment by sindresorhus** [comment]
Sorry, I just realized I missed the notification for this.

**Comment by sindresorhus** [comment]
> Timestamps

I think timestamp can be useful for comparing how far from eachother certain events are. I agree, the date is not very useful, but it should follow the ISO9601 minus the date, so `hours:minutes:seconds.milliseconds`. I think timestamps should be on by default. We can add an option to remove based on future requests.

> Verbosity filter

I'm not sold on this. I think it's better to just be verbose by default and we can add modes to make it less verbose when users requests it from real-world experience.

> Static logger function

Yeah, it should be on the instance.


---

I think it should be `{debug: {name: 'foo'}}`, not `{debugName: 'foo'}`. The former lets us add more debug-related options later on.

---

The docs and TS docs comments needs more work regarding formatting and typos.

**Comment by sbencoding** [comment]
Here are some of the changes I've made:
* `name` for debugging is passed as part of an object that contains debugging setup data in the constructor
* `debugLogger` is no longer static, instead `logger` is set in the debug options
* `isDebug` that was on the instance is now set in debug options as well. The global `isDebug` that controls debugging for all instances is kept as it was before
* Add timestamp to the default logger function
* Update tests
* Update documentation (both readme and typescript)

Let me know if I've missed anything, or there's something that could be done in a better way

P.S. finally I pass all the checks yay!

File: index.d.ts [hunk 3]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
 declare const listenerRemoved: unique symbol;
 type OmnipresentEventData = {[listenerAdded]: Emittery.ListenerChangedData; [listenerRemoved]: Emittery.ListenerChangedData};
 
+/**
+Emittery can collect and log debug information.
+
+To enable this feature set the DEBUG environment variable to 'emittery' or '*'. Additionally you can set the static `isDebug` variable to true on the Emittery class, or `myEmitter.debug.enabled` on an instance of it for debugging a single instance.
+
+See API for more information on how debugging works
```
Comment (sindresorhus) [inline index.d.ts:22]: All sentences should end with a `.`

File: index.d.ts [hunk 4]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+*/
+type DebugLogger = (type: string, debugName: string, eventName?: EventName, eventData?: Record<string, any>) => void;
```
Comment (sindresorhus) [inline index.d.ts:24]: The event data here should be strongly-typed.

File: index.d.ts [hunk 5]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+
+	@default Empty string
+	@example
+	```
+	const Emittery = require('emittery');
+	Emittery.isDebug = true;
+	const emitter = new Emittery({debug: {name: 'myEmitter'}});
+	emitter.on('test', data => { // do something });
+	//=> [16:43:20.417][emittery:subscribe][myEmitter] Event Name: test
+	//	data: undefined
```
Comment (sindresorhus) [inline index.d.ts:41]: This is too cramped. Use more empty lines to make it more readable. The example should also include the part that emits the event. Generally, code examples should be runnable on their own.

Applies to all code examples.

File: index.d.ts [hunk 6]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+type DebugLogger = (type: string, debugName: string, eventName?: EventName, eventData?: Record<string, any>) => void;
+
+/**
+Configure debug options of an instance
+*/
+interface DebugOptions {
+	/**
+	Define a name for the instance of Emittery to use when outputting debug data.
+
+	@default Empty string
```
Comment (sindresorhus) [inline index.d.ts:33]: This is not a valid value for `@default`.

File: index.d.ts [hunk 7]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+	```
+	const Emittery = require('emittery');
+	Emittery.isDebug = true;
+	const emitter = new Emittery({debug: {name: 'myEmitter'}});
+	emitter.on('test', data => { // do something });
+	//=> [16:43:20.417][emittery:subscribe][myEmitter] Event Name: test
+	//	data: undefined
+	```
+	*/
+	name: string;
```
Comment (sindresorhus) [inline index.d.ts:44]:
```suggestion
	readonly name: string;
```

File: index.d.ts [hunk 8]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+
+	/**
+	Enable/Disable debug logging just for this instance
```
Comment (sindresorhus) [inline index.d.ts:47]:
```suggestion
	Toggle debug logging just for this instance.
```

File: index.d.ts [hunk 9]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+
+/**
+Configure debug options of an instance
+*/
+interface DebugOptions {
+	/**
+	Define a name for the instance of Emittery to use when outputting debug data.
+
+	@default Empty string
+	@example
```
Comment (sindresorhus) [inline index.d.ts:34]:
```suggestion

	@example
```

File: index.d.ts [hunk 10]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+	enabled?: boolean;
+
+	/**
+	Function that handles debug data
+
+	@default
+	```
+	(type, debugName, eventName, eventData) => {
+		if (typeof eventData === 'object') {
+			eventData = JSON.stringify(eventData);
```
Comment (sindresorhus) [inline index.d.ts:69]: I think it's easier to just always stringily the value.

File: index.d.ts [hunk 11]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+		}
+
+		if (typeof eventName === 'symbol') {
+			eventName = eventName.toString();
+		}
```
Comment (sindresorhus) [inline index.d.ts:74]: This is moot. Template string interpolation already handles this.

File: index.d.ts [hunk 12]
```diff
@@ -14,6 +14,97 @@ declare const listenerAdded: unique symbol;
...
+	emitter.on('test', data => { // do something });
+
+	//=> [subscribe]: test
+	```
+	*/
+	logger?: DebugLogger;
+}
+
+/**
+Configure a new instance of Emittery
```
Comment (sindresorhus) [inline index.d.ts:102]: Incorrect description.

File: index.d.ts [hunk 13]
```diff
@@ -104,6 +217,39 @@ declare class Emittery<
...
+				eventName = eventName.toString();
+			}
+
+			const currentTime = new Date();
+			const logTime = `${currentTime.getHours()}:${currentTime.getMinutes()}:${currentTime.getSeconds()}.${currentTime.getMilliseconds()}`;
+			console.log(`[${logTime}][emittery:${type}][${debugName}] Event Name: ${eventName}\n\tdata: ${eventData}`);
+		}
+	}
+	```
+	*/
```
Comment (sindresorhus) [inline index.d.ts:243]: This `@default` block here is not needed. It's already document on the individual options.

File: index.js [hunk 14]
```diff
@@ -9,6 +9,8 @@ const resolvedPromise = Promise.resolve();
 const listenerAdded = Symbol('listenerAdded');
 const listenerRemoved = Symbol('listenerRemoved');
 
+let globalDebugFlag = false;
```
Comment (sindresorhus) [inline index.js:12]:
```suggestion
let isGlobalDebugEnabled = false;
```

File: index.js [hunk 15]
```diff
@@ -189,10 +191,42 @@ class Emittery {
 		};
 	}
 
-	constructor() {
+	static get isDebug() {
```
Comment (sindresorhus) [inline index.js:194]:
```suggestion
	static get isDebugEnabled() {
```

Just to make it clear.

File: index.js [hunk 16]
```diff
@@ -189,10 +191,42 @@ class Emittery {
...
 		eventsMap.set(this, new Map());
 		producersMap.set(this, new Map());
+		this.debug = options.debug || {};
+		if (!this.debug.name) {
+			this.debug.name = '';
+		}
+
+		if (!this.debug.enabled) {
+			this.debug.enabled = false;
+		}
```
Comment (sindresorhus) [inline index.js:213]: I don't get what the point of this is.

File: index.js [hunk 17]
```diff
@@ -219,6 +257,10 @@ class Emittery {
 			assertEventName(eventName);
 			getListeners(this, eventName).delete(listener);
 
+			if (Emittery.isDebug || this.debug.enabled) {
+				this.debug.logger('unsubscribe', this.debug.name, eventName, undefined);
+			}
```
Comment (sindresorhus) [inline index.js:262]: This is repeated many times. There should be a utility method for this.

File: readme.md [hunk 18]
```diff
@@ -44,7 +44,116 @@ Emittery accepts strings and symbols as event names.
 
 Symbol event names can be used to avoid name collisions when your classes are extended, especially for internal events.
 
-### emitter = new Emittery()
+### isDebug
+Enables/Disables debug mode for all instances
```
Comment (sindresorhus) [inline readme.md:48]:
```suggestion
### isDebug

Toggle debug mode for all instances.
```

Formatting

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
The code looks good, but the docs formatting and text needs more work.

**Comment by sbencoding** [comment]
I implemented the requested changes
1. I wasn't sure how to make `DebugLogger` in `index.d.ts` strongly-typed properly, but I've tried, please let me know if there's a better way to do this, or if how I did it is correct at all.
2. I couldn't remove the following code
```js
if (typeof eventName === 'symbol') {
        eventName = eventName.toString();
}
```
Because it throws `TypeError: Cannot convert a Symbol value to a string` when the string interpolation happens.
I've tried a small, isolated example:
```js
let s = Symbol('test');
console.log(`Symbol to string: ${s}`);
```
This results in the same error, so I don't know if there's a way to work around this.

File: readme.md [hunk 19]
```diff
@@ -56,41 +57,50 @@ Emittery.isDebug = true;
...
 
-Configure the debugging options for this instance
+Configure the debugging options for this instance.
 
 ###### name
 
 Type: `string`
 
-Default: `Empty string`
+Default: `null`
```
Comment (sindresorhus) [inline readme.md]: I prefer using `undefined` over `null`.

File: readme.md [hunk 20]
```diff
@@ -149,7 +165,9 @@ const emitter = new Emittery({
 		logger: myLogger
 	}
 });
+
 emitter.on('test', data => { // do something });
+emitter.emit('test');
```
Comment (sindresorhus) [inline readme.md:168]: Readme and index.d.ts should be in sync.

**Comment by sindresorhus** [comment]
> I wasn't sure how to make DebugLogger in index.d.ts strongly-typed properly, but I've tried, please let me know if there's a better way to do this, or if how I did it is correct at all.

I would recommend writing some type tests to confirm it's correct: https://github.com/sindresorhus/emittery/blob/main/index.test-d.ts

> I couldn't remove the following code

Ah yeah. I forgot it's not interpolatable.

**Comment by sbencoding** [comment]
I've implemented the changes, and wrote some type tests for the strongly-typed logger

**Pull request merged**